drivers/sensor/sun8i-thermal.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright © 2016, ARM Limited and Contributors. All rights reserved.
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <debug.h>
+#include <error.h>
+#include <limits.h>
+#include <mmio.h>
+#include <sensor.h>
+#include <sensor/sun8i-thermal.h>
+
+#define THS_0_DATA_REG         0x80
+#define THS_1_DATA_REG         0x84
+#define THS_2_DATA_REG         0x88
+
+#define THS_CTL_REG0           0x00
+#define THS_CTL_REG1           0x04
+#define THS_CTL_REG2           0x40
+
+#define THS_FILTER_CONTROL_REG 0x70
+
+struct sun8i_thermal_sensor_info {
+	struct   sensor_info info;
+	uint16_t             address;
+};
+
+enum {
+	THS0,
+	THS1,
+	THS2,
+	SENSOR_COUNT
+};
+
+static struct sun8i_thermal_sensor_info thermal_sensors[SENSOR_COUNT] = {
+	[THS0] = {
+		.info = {
+			.name       = "ths0",
+			.offset     = 680,
+			.multiplier = -856,
+		},
+		.address = 0x80,
+	},
+	[THS1] = {
+		.info = {
+			.name       = "ths1",
+			.offset     = 680,
+			.multiplier = -856,
+		},
+		.address = 0x84,
+	},
+	[THS2] = {
+		.info = {
+			.name       = "ths2",
+			.offset     = 680,
+			.multiplier = -856,
+		},
+		.address = 0x88,
+	},
+};
+
+static struct sensor_info *
+sun8i_thermal_get_info(struct device *dev __unused, uint8_t id)
+{
+	assert(id < SENSOR_COUNT);
+
+	return &thermal_sensors[id].info;
+}
+
+static int
+sun8i_thermal_read_raw(struct device *dev, uint8_t id, uint32_t *raw)
+{
+	uint16_t addr = thermal_sensors[id].address;
+
+	*raw = mmio_read32(dev->regs + addr) & BITMASK(0, 11);
+
+	return SUCCESS;
+}
+
+static int
+sun8i_thermal_probe(struct device *dev)
+{
+	uint32_t reg;
+	int err;
+
+	/* Set up two clocks for the thermal sensor */
+	if ((err = dm_setup_clocks(dev, 2)))
+		return err;
+
+	/* Start calibration. */
+	mmio_write32(dev->regs + THS_CTL_REG1, BIT(17));
+	while (mmio_read32(dev->regs + THS_CTL_REG1) & BIT(17)) {
+		/* Wait for for calibration to clear */
+	}
+
+	/* Set aquire times. */
+	mmio_write32(dev->regs + THS_CTL_REG0, 0x190);
+	mmio_write32(dev->regs + THS_CTL_REG2, 0x190 << 16);
+
+	/* Enable filter, average over 8 values. */
+	mmio_write32(dev->regs + THS_FILTER_CONTROL_REG, 0x06);
+
+	/* Enable sensor measurement for the CPU sensors and both GPU sensor. *
```

Comment:
"Enable ~sensor~" to avoid wrapping. Also "GPU sensor_s_"

Reply:
Done